compiler/src/iree/compiler/GlobalOptimization/FuseHorizontalContractions.cpp

```diff
@@ -0,0 +1,424 @@
+#include "iree/compiler/GlobalOptimization/Utils.h"
+#include "mlir/IR/Dominance.h"
+
+#include "mlir/Dialect/Arith/IR/Arith.h"
+#include "mlir/Dialect/Linalg/IR/Linalg.h"
+#include "mlir/Dialect/Linalg/IR/LinalgInterfaces.h"
+#include "mlir/Dialect/Linalg/Transforms/Transforms.h"
+#include "mlir/Dialect/Math/IR/Math.h"
+#include "mlir/Dialect/Tensor/IR/Tensor.h"
+#include "mlir/Dialect/Tensor/Transforms/Transforms.h"
+#include "mlir/Dialect/Tensor/Utils/Utils.h"
+#include "mlir/Dialect/Utils/ReshapeOpsUtils.h"
+#include "mlir/Dialect/Utils/StructuredOpsUtils.h"
+#include "mlir/IR/BuiltinTypes.h"
+#include "mlir/IR/OpDefinition.h"
+#include "mlir/IR/PatternMatch.h"
+#include "mlir/Pass/Pass.h"
+#include "mlir/Support/LLVM.h"
+#include "mlir/Support/LogicalResult.h"
+#include "mlir/Transforms/GreedyPatternRewriteDriver.h"
+
+#define DEBUG_TYPE "iree-global-opt-fuse-horizontal-contraction"
+#define DBGS() (llvm::dbgs() << '[' << DEBUG_TYPE << "] ")
+
+using namespace mlir;
+
+namespace mlir::iree_compiler::GlobalOptimization {
+
+namespace {
+
+struct FuseHorizontalContractionsPass
+    : public FuseHorizontalContractionsBase<FuseHorizontalContractionsPass> {
+
+  void getDependentDialects(DialectRegistry &registry) const override {
+    registry.insert<linalg::LinalgDialect, IREE::Flow::FlowDialect,
+                    math::MathDialect>();
+  }
+  FuseHorizontalContractionsPass() {}
+  FuseHorizontalContractionsPass(const FuseHorizontalContractionsPass &pass)
+      : FuseHorizontalContractionsPass() {}
+
+  void runOnOperation() override;
+};
+
+} // namespace
+
+/// Structs that captures the ops that are to be fused
+struct HorizontalFusionGroup {
+  // Contractions op that are to be fused.
+  SmallVector<linalg::LinalgOp> contractionOps;
+  // Optional truncate operations that could be following the contraction op.
+  std::optional<SmallVector<linalg::GenericOp>> truncateOps;
+  // Operation that dominates all the ops of the group.
+  Operation *dominatedByAll;
+};
+
+/// Get user of operation that is a truncate operation.
+static std::optional<linalg::GenericOp> getTruncUser(Operation *op) {
+  if (op->getNumResults() != 1) {
+    return std::nullopt;
+  }
+  Value result = op->getResult(0);
+  if (!result.hasOneUse()) {
+    return std::nullopt;
+  }
+  Operation *user = *result.user_begin();
+  auto genericOp = dyn_cast<linalg::GenericOp>(user);
+  if (!genericOp) {
+    return std::nullopt;
+  }
+  if (genericOp.getNumDpsInputs() != 1 || genericOp.getNumDpsInits() != 1) {
+    return std::nullopt;
+  }
+  if (llvm::any_of(genericOp.getIndexingMapsArray(),
+                   [](AffineMap map) { return !map.isIdentity(); })) {
+    return std::nullopt;
+  }
+  if (genericOp.getNumParallelLoops() != genericOp.getNumLoops()) {
+    return std::nullopt;
+  }
+  auto yieldOp = cast<linalg::YieldOp>(genericOp.getBody()->getTerminator());
+  auto yieldDefOp = yieldOp->getOperand(0).getDefiningOp<arith::TruncFOp>();
```

Comment:
Do we plan to add `arith::TruncIOp` to the allow list? If not, maybe we can rename the function to `getTruncFUser`